### PR TITLE
Update soil layer data erosion attributes from percents to fractions

### LINKED ISSUE #467: Rename and refactor Erosion attributes in `LayerData`
# Problem
The set of attributes (`percent_organic_carbon_content`, `percent_clay_content`, `percent_sand_content`, `percent_silt_content`, and `percent_rock_content`) used by the erosion submodule in the soil module and located in `LayerData` follow the patterns:
1. The name takes the form of `percent_<name>_content`
2. The value represents a percent, and is in the range [0, 100] (1% = 1)
This is an issue because it is not consistent with other modules, and can lead to confusion.

# Solution
Refactor this set of attributes so that they follow the patterns:
1. [ ] The name takes the form of `<name>_proportion_content` (or whatever pattern; as long as it is clear, simple, and consistently used the naming pattern doesn't really matter)
2. [ ] The value should be changed to be in the range [0.0, 1.0] (So that 1% = 0.01)

After these patterns have been implemented, the following tasks need to be accomplished:
- [ ] All docstrings of these attributes need to be refactored indicate clearly what they mean
- [ ] All methods that use these attributes are refactored to adjust for their change in value (and also have their docstrings updated to indicate the scale of the value they take and how they adjust it)

### PR DESCRIPTION
**Pull request by ew3361zh**
Renames and refactors Erosion attributes in LayerData as well as everywhere they're referenced.

### Context

Issue(s) closed by this pull request: closes #467

- [x] The [changelog](https://docs.google.com/spreadsheets/d/1U-3igxdstHTFb6k5dVcE5-x9IL0r6q18eUiwwytKZyA/edit#gid=0) has been updated.

### What

The set of attributes used by the erosion submodule in the soil module and located in LayerData have been updated from percents to fractions:
`percent_organic_carbon_content` -> `organic_carbon_fraction`
`percent_clay_content` -> `clay_fraction`
`percent_sand_content` -> `sand_fraction`
`percent_silt_content` -> `silt_fraction`
`percent_rock_content` -> `rock_fraction`

The previous format of the attribute was to be as a percent (e.g. 55.2%) and the new fraction version would be (0.552).

All references (in the model, docstrings, in testing) have been updated to reflect these changes as well.

### Why

This is done to be consistent with other parts of the C&S Module as well as other parts of the RuFaS model.

### How

Using search and update for all the references.
Adjusting the equations, references, and docstrings accordingly.
Updating test cases to match new format.

### Test plan

Ran successfully with all metadata scenarios. Testing updated and passing as before.

### CODE REVIEW

File: RUFAS/routines/field/soil/nitrogen_cycling/denitrification.py [hunk 1]
```diff
@@ -82,8 +83,8 @@ def _calculate_denitrification_amount(
             Rate coefficient that regulates denitrification in this layer of soil (unitless)
         temp_factor : float
             The nutrient cycling temperature factor of this soil layer (unitless)
-        percent_organic_carbon_content : float
-            The percent of this soil layer that is made up of organic carbon, in the range [0, 100] (unitless)
+        organic_carbon_fraction : float
+            The fraction of this soil layer that is made up of organic carbon, in the range [0, 1.0] (unitless)
```
Comment (matthew7838) [inline RUFAS/routines/field/soil/nitrogen_cycling/denitrification.py]:
```suggestion
            The fraction of this soil layer that is made up of organic carbon, in the range [0, 1.0] (unitless).
```

File: RUFAS/output_manager.py [hunk 2]
```diff
@@ -549,25 +549,86 @@ def _dict_to_csv_column_list(self, variable_name: str, data_dict: Dict[str, List
...
+    ) -> str:
+        """Get the units substring for a column title.
+
+        Parameters
+        ----------
+        variable_name : str
+            The name of the variable or group of variables associated with the units.
+        units : str | Dict[str, str] | None
+            The units associated with the data.
+        subkey : str | None, optional
```
Comment (matthew7838) [inline RUFAS/output_manager.py]:
```suggestion
        units : str | Dict[str, str] | None
            The units associated with the data.
        subkey : str, optional, default None
```

Comment (matthew7838) [reply RUFAS/output_manager.py]: Is the unit a string or a dictionary?

File: RUFAS/routines/field/soil/nitrogen_cycling/denitrification.py [hunk 3]
```diff
@@ -82,8 +83,8 @@ def _calculate_denitrification_amount(
             Rate coefficient that regulates denitrification in this layer of soil (unitless)
         temp_factor : float
             The nutrient cycling temperature factor of this soil layer (unitless)
```
Comment (matthew7838) [inline RUFAS/routines/field/soil/nitrogen_cycling/denitrification.py]: This should be something that I missed while reformatting it. They're all missing periods.

File: RUFAS/output_manager.py [hunk 4]
```diff
@@ -549,25 +549,86 @@ def _dict_to_csv_column_list(self, variable_name: str, data_dict: Dict[str, List
...
-                for subkey in csv_column_lists.keys():
-                    column_title = f"{variable_name}.{subkey}"
-                    column_list.append(pd.Series(csv_column_lists[subkey], dtype=object, name=column_title))
-            else:
-                column_title = f"{variable_name}"
-                column_list.append(pd.Series(data_list, dtype=object, name=column_title))
+        units = data_dict["info_maps"][0]["units"] if data_dict.get("info_maps", []) else None
+        data_list = data_dict["values"]
+        if data_list and isinstance(data_list[0], dict):
+            csv_column_lists: Dict[str, List[Any]] = {subkey: [] for item in data_list for subkey in item.keys()}
```
Comment (matthew7838) [inline RUFAS/output_manager.py]:
```suggestion
            csv_column_lists: Dict[str, List[Any]] = {
            subkey: [] 
            for item in data_list 
            for subkey in item.keys()
}
```
More readable if this is written in multiple lines.

**Review by matthew7838 — APPROVED** [review]
LGTM, there are some undocumented changes in the templates. If those were meant to be changed then it's all good. Great work Niko.

File: RUFAS/output_manager.py (hunk 4 above)
Comment (ew3361zh) [reply RUFAS/output_manager.py]: Ugh, this unfortunately isn't my code. It looks like I may have accidentally rebased incorrectly and pulled in changes and commits from another PR.

File: input/data/soil/default_soil.json [hunk 5]
```diff
@@ -64,11 +64,11 @@
             "saturated_hydraulic_conductivity": 20,
             "initial_temperature": 13.38623,
             "bulk_density": 1.3,
-            "percent_organic_carbon_content": 1.2,
-            "percent_clay_content": 20,
-            "percent_silt_content": 65,
-            "percent_sand_content": 15,
-            "percent_rock_content": 0.0,
+            "organic_carbon_fraction": 0.12,
```
Comment (Isgotkowitz) [inline input/data/soil/default_soil.json]:
```suggestion
            "organic_carbon_fraction": 0.012,
```

File: input/data/soil/default_soil_2.json [hunk 6]
```diff
@@ -42,11 +42,11 @@
             "saturated_hydraulic_conductivity": 9.17,
             "initial_temperature": 14.50797297,
             "bulk_density": 1.42,
-            "percent_organic_carbon_content": 0.348,
-            "percent_clay_content": 27.27,
-            "percent_silt_content": 59.09,
-            "percent_sand_content": 13.64,
-            "percent_rock_content": 0.0,
+            "organic_carbon_fraction": 0.000348,
```
Comment (Isgotkowitz) [inline input/data/soil/default_soil_2.json]:
```suggestion
            "organic_carbon_fraction": 0.00348,
```

File: input/data/soil/default_soil_2.json [hunk 7]
```diff
@@ -66,11 +66,11 @@
             "saturated_hydraulic_conductivity": 23.29,
             "initial_temperature": 13.38623,
             "bulk_density": 1.6,
-            "percent_organic_carbon_content": 0.145,
-            "percent_clay_content": 22.33,
-            "percent_silt_content": 63.11,
-            "percent_sand_content": 14.56,
-            "percent_rock_content": 0.0,
+            "organic_carbon_fraction": 0.000145,
```
Comment (Isgotkowitz) [inline input/data/soil/default_soil_2.json]:
```suggestion
            "organic_carbon_fraction": 0.00145,
```

File: input/data/soil/default_soil_2.json [hunk 8]
```diff
@@ -90,11 +90,11 @@
             "saturated_hydraulic_conductivity": 23.29,
             "initial_temperature": 13.38623,
             "bulk_density": 1.5,
-            "percent_organic_carbon_content": 0.145,
-            "percent_clay_content": 13.04,
-            "percent_silt_content": 70.66,
-            "percent_sand_content": 16.30,
-            "percent_rock_content": 0.0,
+            "organic_carbon_fraction": 0.000145,
```
Comment (Isgotkowitz) [inline input/data/soil/default_soil_2.json]:
```suggestion
            "organic_carbon_fraction": 0.00145,
```

File: input/metadata/properties/default.json [hunk 9]
```diff
@@ -4698,40 +4698,40 @@
...
-        "percent_organic_carbon_content": {
+        "organic_carbon_fraction": {
           "type": "number",
-          "description": "Percent of soil weight that is organic carbon.\nUnitless.\nReference: SWAT Input .SOL - SOL_CBN.",
+          "description": "Fraction of soil weight that is organic carbon.\nUnitless.\nReference: SWAT Input .SOL - SOL_CBN.",
           "minimum": 0.0,
-          "maximum": 100.0,
-          "default": 1.2
+          "maximum": 1.0,
+          "default": 0.12
```
Comment (Isgotkowitz) [inline input/metadata/properties/default.json]:
```suggestion
          "default": 0.012
```

File: tests/soil_crop_tests/soil_tests/test_soil_erosion.py [hunk 10]
```diff
@@ -206,17 +206,17 @@ def test_determine_topographic_factor(length: float, avg_slope: float) -> None:
...
-        0.0492184949,
-        0.10495492330,
+        0.005,
+        0.002194,
+        0.0019493,
+        0.00492184949,
+        0.010495492330,
     ],
 )
 def test_determine_coarse_fragment_factor(percent_rock: float) -> None:
```
Comment (Isgotkowitz) [inline tests/soil_crop_tests/soil_tests/test_soil_erosion.py]: Maybe rename this to rock_fraction

File: tests/soil_crop_tests/manager_tests/test_field_manager.py [hunk 11]
```diff
@@ -1616,11 +1616,11 @@ def test_setup_soil_layer_error(config: Dict) -> None:
                     "saturated_hydraulic_conductivity": 20,
                     "initial_temperature": 13.38623,
                     "bulk_density": 1.3,
-                    "percent_organic_carbon_content": 0.012,
-                    "percent_clay_content": 20,
-                    "percent_silt_content": 65,
-                    "percent_sand_content": 15,
-                    "percent_rock_content": 0.0,
+                    "organic_carbon_fraction": 0.012,
+                    "clay_fraction": 20,
```
Comment (Isgotkowitz) [inline tests/soil_crop_tests/manager_tests/test_field_manager.py]:
```suggestion
                    "clay_fraction": 0.20,
```

File: tests/soil_crop_tests/manager_tests/test_field_manager.py [hunk 12]
```diff
@@ -1592,11 +1592,11 @@ def test_setup_soil_layer_error(config: Dict) -> None:
                     "saturated_hydraulic_conductivity": 20,
                     "initial_temperature": 14.50797297,
                     "bulk_density": 1.3,
-                    "percent_organic_carbon_content": 0.012,
-                    "percent_clay_content": 20,
-                    "percent_silt_content": 65,
-                    "percent_sand_content": 15,
-                    "percent_rock_content": 0.0,
+                    "organic_carbon_fraction": 0.012,
+                    "clay_fraction": 20,
```
Comment (Isgotkowitz) [inline tests/soil_crop_tests/manager_tests/test_field_manager.py]:
```suggestion
                    "clay_fraction": 0.20,
```

File: tests/soil_crop_tests/manager_tests/test_field_manager.py [hunk 13]
```diff
@@ -1568,11 +1568,11 @@ def test_setup_soil_layer_error(config: Dict) -> None:
                     "saturated_hydraulic_conductivity": 20,
                     "initial_temperature": 15.77575,
                     "bulk_density": 1.3,
-                    "percent_organic_carbon_content": 0.012,
-                    "percent_clay_content": 20,
-                    "percent_silt_content": 65,
-                    "percent_sand_content": 15,
-                    "percent_rock_content": 0.0,
+                    "organic_carbon_fraction": 0.012,
+                    "clay_fraction": 20,
```
Comment (Isgotkowitz) [inline tests/soil_crop_tests/manager_tests/test_field_manager.py]:
```suggestion
                    "clay_fraction": 0.20,
```

File: tests/soil_crop_tests/manager_tests/test_field_manager.py [hunk 14]
```diff
@@ -1531,11 +1531,11 @@ def test_setup_soil_layer_error(config: Dict) -> None:
                     "saturated_hydraulic_conductivity": 23.29,
                     "initial_temperature": 13.38623,
                     "bulk_density": 1.5,
-                    "percent_organic_carbon_content": 0.012,
-                    "percent_clay_content": 13.04,
-                    "percent_silt_content": 70.66,
-                    "percent_sand_content": 16.30,
-                    "percent_rock_content": 0.0,
+                    "organic_carbon_fraction": 0.012,
+                    "clay_fraction": 13.04,
```
Comment (Isgotkowitz) [inline tests/soil_crop_tests/manager_tests/test_field_manager.py]:
```suggestion
                    "clay_fraction": 0.1304,
```

File: tests/soil_crop_tests/manager_tests/test_field_manager.py [hunk 15]
```diff
@@ -1506,11 +1506,11 @@ def test_setup_soil_layer_error(config: Dict) -> None:
                     "saturated_hydraulic_conductivity": 23.29,
                     "initial_temperature": 13.38623,
                     "bulk_density": 1.6,
-                    "percent_organic_carbon_content": 0.012,
-                    "percent_clay_content": 22.33,
-                    "percent_silt_content": 63.11,
-                    "percent_sand_content": 14.56,
-                    "percent_rock_content": 0.0,
+                    "organic_carbon_fraction": 0.012,
+                    "clay_fraction": 22.33,
```
Comment (Isgotkowitz) [inline tests/soil_crop_tests/manager_tests/test_field_manager.py]:
```suggestion
                    "clay_fraction": 0.2233,
```

File: tests/soil_crop_tests/manager_tests/test_field_manager.py [hunk 16]
```diff
@@ -1481,11 +1481,11 @@ def test_setup_soil_layer_error(config: Dict) -> None:
                     "saturated_hydraulic_conductivity": 9.17,
                     "initial_temperature": 14.50797297,
                     "bulk_density": 1.42,
-                    "percent_organic_carbon_content": 0.012,
-                    "percent_clay_content": 27.27,
-                    "percent_silt_content": 59.09,
-                    "percent_sand_content": 13.64,
-                    "percent_rock_content": 0.0,
+                    "organic_carbon_fraction": 0.012,
+                    "clay_fraction": 27.27,
```
Comment (Isgotkowitz) [inline tests/soil_crop_tests/manager_tests/test_field_manager.py]:
```suggestion
                    "clay_fraction": 0.2727,
```

File: tests/soil_crop_tests/manager_tests/test_field_manager.py [hunk 17]
```diff
@@ -1456,11 +1456,11 @@ def test_setup_soil_layer_error(config: Dict) -> None:
                     "saturated_hydraulic_conductivity": 9.17,
                     "initial_temperature": 15.77575,
                     "bulk_density": 1.34,
-                    "percent_organic_carbon_content": 0.012,
-                    "percent_clay_content": 21.95,
-                    "percent_silt_content": 63.42,
-                    "percent_sand_content": 14.63,
-                    "percent_rock_content": 0.0,
+                    "organic_carbon_fraction": 0.012,
+                    "clay_fraction": 21.95,
```
Comment (Isgotkowitz) [inline tests/soil_crop_tests/manager_tests/test_field_manager.py]:
```suggestion
                    "clay_fraction": 0.2195,
```

File: tests/soil_crop_tests/manager_tests/test_field_manager.py [hunk 18]
```diff
@@ -1379,20 +1379,20 @@ def test_crop_schedule_setup(
                 field_capacity_water_concentration=0.306,
                 saturation_point_water_concentration=0.5,
                 saturated_hydraulic_conductivity=9.17,
-                percent_clay_content=27.27,
+                clay_fraction=27.27,
```
Comment (Isgotkowitz) [inline tests/soil_crop_tests/manager_tests/test_field_manager.py]:
```suggestion
                clay_fraction=0.2727,
```

File: tests/soil_crop_tests/manager_tests/test_field_manager.py [hunk 19]
```diff
@@ -1355,11 +1355,11 @@ def test_crop_schedule_setup(
                 "saturated_hydraulic_conductivity": 9.17,
                 "initial_temperature": 14.50797297,
                 "bulk_density": 1.42,
-                "percent_organic_carbon_content": 0.012,
-                "percent_clay_content": 27.27,
-                "percent_silt_content": 59.09,
-                "percent_sand_content": 13.64,
-                "percent_rock_content": 0.0,
+                "organic_carbon_fraction": 0.012,
+                "clay_fraction": 27.27,
```
Comment (Isgotkowitz) [inline tests/soil_crop_tests/manager_tests/test_field_manager.py]:
```suggestion
                "clay_fraction": 0.2727,
```

File: tests/soil_crop_tests/manager_tests/test_field_manager.py [hunk 20]
```diff
@@ -1324,20 +1324,20 @@ def test_crop_schedule_setup(
                 field_capacity_water_concentration=0.29,
                 saturation_point_water_concentration=0.58,
                 saturated_hydraulic_conductivity=9.17,
-                percent_clay_content=21.95,
+                clay_fraction=21.95,
```
Comment (Isgotkowitz) [inline tests/soil_crop_tests/manager_tests/test_field_manager.py]:
```suggestion
                clay_fraction=0.2195,
```

File: tests/soil_crop_tests/manager_tests/test_field_manager.py [hunk 21]
```diff
@@ -1300,11 +1300,11 @@ def test_crop_schedule_setup(
                 "saturated_hydraulic_conductivity": 9.17,
                 "initial_temperature": 15.77575,
                 "bulk_density": 1.34,
-                "percent_organic_carbon_content": 0.012,
-                "percent_clay_content": 21.95,
-                "percent_silt_content": 63.42,
-                "percent_sand_content": 14.63,
-                "percent_rock_content": 0.0,
+                "organic_carbon_fraction": 0.012,
+                "clay_fraction": 21.95,
```
Comment (Isgotkowitz) [inline tests/soil_crop_tests/manager_tests/test_field_manager.py]:
```suggestion
                "clay_fraction": 0.2195,
```

**Review by Isgotkowitz — APPROVED** [review]
Noticed some spots in inputs and tests where the percentage had not been correctly converted to a fraction, should be a simple fix. Am commenting for right now, I want to do some more extensive comparison between this branch and main to make sure there aren't any unreasonable changes in erosion amounts or carbon levels. I will come back and approve once I do this.

File: RUFAS/routines/field/soil/soil_erosion.py [hunk 22]
```diff
@@ -225,29 +226,29 @@ def _determine_high_sand_factor(percent_sand_content: float) -> float:
         SWAT Theoretical documentation eqn. 4:1.1.9
 
         """
-        not_sand_fraction = 1 - (percent_sand_content / 100)
+        not_sand_fraction = 1 - (sand_fraction)
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/soil/soil_erosion.py]:
```suggestion
        not_sand_fraction = 1 - sand_fraction
```

File: RUFAS/routines/field/soil/soil_erosion.py [hunk 23]
```diff
@@ -167,9 +168,9 @@ def _determine_clay_silt_ratio_factor(percent_silt_content: float, percent_clay_
         SWAT Theoretical documentation eqn. 4:1.1.7
 
         """
-        if percent_silt_content == 0 and percent_clay_content == 0:
+        if silt_fraction == 0 and clay_fraction == 0:
             raise ValueError("Cannot have percent silt content and percent clay content both be 0")
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/soil/soil_erosion.py]:
```suggestion
            raise ValueError("Cannot have fractions of silt and clay in the soil both be 0")
```

File: RUFAS/routines/field/soil/soil_erosion.py [hunk 24]
```diff
@@ -167,9 +168,9 @@ def _determine_clay_silt_ratio_factor(percent_silt_content: float, percent_clay_
...
-        return (percent_silt_content / (percent_clay_content + percent_silt_content)) ** 0.3
+        return (silt_fraction / ((clay_fraction) + (silt_fraction))) ** 0.3
 
     @staticmethod
     def _determine_carbon_content_factor(percent_organic_carbon: float) -> float:
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/soil/soil_erosion.py]: This method still needs to be updated to account for the switch to fractions.

**Review by Isgotkowitz — COMMENTED** [review]
Found a few places that need to be updated to account of the use of fractions, and not percents. Also, am still seeing minor differences in the default between this branch and main, looking with this filter:
```
{
    "multiple": [
        {
            "name": "Carbon",
            "filters": [
                "FieldDataReporter.send_daily_variables.profile_carbon_total.field='field'"
            ],
            "vertical_aggregation": "average"
        },
        {
            "name": "Erosion",
            "filters": [
                "FieldDataReporter.send_daily_variables.eroded_sediment.field='field'"
            ],
            "vertical_aggregation": "average"
        }
    ]
}
```
I think it would be good to get absolutely no difference between all scenarios before merging, or at least make sure the differences between all scenarios are really small (like >1%)

**Comment by krb78** [comment]
After scouring the code and looking for any reasons there might be these slight differences when changing from % to fraction, I cannot identify any reasons for the differences in soil carbon. Given the differences are so minor (<1% at the end of the entire simulation), I think we are ok to accept this as a rounding difference in the model. While we do work for the highest level of accuracy possible, even the best carbon soil researchers don't have confidence in their measured data as precisely as 20kg/ha (the margins of error are bigger than that).

**Pull request merged**